rasa/src/Rasa/Internal/ActionMonads.hs

```diff
@@ -0,0 +1,157 @@
+{-# language DeriveFunctor
+  , MultiParamTypeClasses
+  , FlexibleInstances
+  , GeneralizedNewtypeDeriving
+  , RankNTypes
+  , ExistentialQuantification
+  , ScopedTypeVariables
+#-}
+module Rasa.Internal.ActionMonads
+  ( Action(..)
+  , BufAction(..)
+  , ActionF(..)
+  , BufActionF(..)
+  , Listener(..)
+  , Listeners
+  , ListenerId(..)
+  , liftActionF
+  , liftBufAction
+  , dispatchEvent
+  ) where
+
+import Rasa.Internal.Editor
+import Rasa.Internal.Buffer
+import Rasa.Internal.Range
+
+import Control.Monad.Free
+import Control.Monad.IO.Class
+
+import Data.Default
+import Data.Typeable
+import qualified Data.Map as M
+
+import qualified Yi.Rope as Y
+
+-- | A wrapper around event listeners so they can be stored in 'Listeners'.
+data Listener = forall a. Listener ListenerId (a -> Action ())
+
+-- | An opaque reverence to a specific registered event-listener.
+-- A ListenerId is used only to remove listeners later with 'Rasa.Internal.Listeners.removeListener'.
+data ListenerId =
+  ListenerId Int TypeRep
+
+instance Eq ListenerId where
+  ListenerId a _ == ListenerId b _ = a == b
+
+-- | A map of Event types to a list of listeners for that event
+type Listeners = M.Map TypeRep [Listener]
+
+
+-- | Free Monad Actions for Action
+data ActionF next =
+    LiftIO (IO next)
+  | forall r. BufferDo [BufRef] (BufAction r) ([r] -> next)
+  | forall event r. Typeable event => AddListener (event -> Action r) (ListenerId -> next)
+  | RemoveListener ListenerId next
+  | forall event. Typeable event => DispatchEvent event next
+  | DispatchActionAsync (IO (Action ())) next
+  | AsyncActionProvider ((Action () -> IO ()) -> IO ()) next
+  | AddBuffer (BufRef -> next)
+  | GetBufRefs ([BufRef] -> next)
+  | forall ext. (Typeable ext, Show ext, Default ext) => GetExt (ext -> next)
+  | forall ext. (Typeable ext, Show ext, Default ext) => SetExt ext next
+  | GetEditor (Editor -> next)
+  | GetBuffer BufRef (Maybe Buffer -> next)
+  | Exit next
+  | ShouldExit (Bool -> next)
```

Comment:
Standalone deriving? You shouldn't have to resort to manually writing an instance unless you want an unlawful functor instance.